## DATASET DETIAL
**VIN (1-10)** : `The Vehicle Identification Number (VIN) is a unique code used to identify individual motor vehicles.`

**County:** `The county where the vehicle is registered.`

**City:** `The city where the vehicle is registered.`

**State:** `The state where the vehicle is registered.`

**Postal Code:** `The postal code for the vehicle's registration location.`

**Model Year:** `The year the vehicle model was manufactured.`

**Make:** `The manufacturer of the vehicle.`

**Model:** `The specific model of the vehicle.`

**Electric Vehicle Type:** `The type of electric vehicle, e.g., Battery Electric Vehicle (BEV), Plug-in Hybrid Electric Vehicle (PHEV).`

**Clean Alternative Fuel Vehicle (CAFV) Eligibility:** `Indicates if the vehicle is eligible for clean alternative fuel incentives.`

**Electric Range:** `The range the electric vehicle can travel on a single charge.`

## Importing Libraries

In [697]:
import os

import numpy as np

import pandas as pd 

import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split

## Reading Data

In [495]:
# Project directory represent the project folder
PROJECT_DIR = r"E:\Electirc Viechels"
# this line of code access data if we have a folder name data
DATA_DIR = "data"

In [496]:
# This function access the data 
def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    return pd.read_csv(file_path)

In [497]:
veichels = get_data('veichel')
veichels

VIN (1-10)        County       City State  Postal Code  Model Year  \
0       5YJSA1E22K          King    Seattle    WA      98112.0        2019   
1       3MW39FS05R        Yakima     Zillah    WA      98953.0        2024   
2       1N4AZ0CP0F          King       Kent    WA      98031.0        2015   
3       5YJSA1H20F     Snohomish    Bothell    WA      98012.0        2015   
4       JTMAB3FV1N        Yakima     Yakima    WA      98908.0        2022   
...            ...           ...        ...   ...          ...         ...   
191402  5YJ3E1EA0K        Pierce     Sumner    WA      98390.0        2019   
191403  JN1BF0BA7P        Pierce   Wilkeson    WA      98396.0        2023   
191404  7SAYGDEE4R          King  Fall City    WA      98024.0        2024   
191405  7SAXCBE65N  Grays Harbor   Aberdeen    WA      98520.0        2022   
191406  7SAYGAEEXP        Benton   Richland    WA      99352.0        2023   

          Make       Model                   Electric Vehicle Type  \
0        TESLA     MODEL S          Battery Electric Vehicle (BEV)   
1          BMW        330E  Plug-in Hybrid Electric Vehicle (PHEV)   
2       NISSAN        LEAF          Battery Electric Vehicle (BEV)   
3        TESLA     MODEL S          Battery Electric Vehicle (BEV)   
4       TOYOTA  RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
...        ...         ...                                     ...   
191402   TESLA     MODEL 3          Battery Electric Vehicle (BEV)   
191403  NISSAN       ARIYA          Battery Electric Vehicle (BEV)   
191404   TESLA     MODEL Y          Battery Electric Vehicle (BEV)   
191405   TESLA     MODEL X          Battery Electric Vehicle (BEV)   
191406   TESLA     MODEL Y          Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0                 Clean Alternative Fuel Vehicle Eligible             270   
1                   Not eligible due to low battery range              20   
2                 Clean Alternative Fuel Vehicle Eligible              84   
3                 Clean Alternative Fuel Vehicle Eligible             208   
4                 Clean Alternative Fuel Vehicle Eligible              42   
...                                                   ...             ...   
191402            Clean Alternative Fuel Vehicle Eligible             220   
191403  Eligibility unknown as battery range has not b...               0   
191404  Eligibility unknown as battery range has not b...               0   
191405  Eligibility unknown as battery range has not b...               0   
191406  Eligibility unknown as battery range has not b...               0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
0               0                  43.0       202233958   
1               0                  15.0       264425178   
2               0                  11.0       114962025   
3               0                   1.0       232724670   
4               0                  14.0       221023589   
...           ...                   ...             ...   
191402          0                  31.0       314927896   
191403          0                  31.0       236213120   
191404          0                   5.0       261335896   
191405          0                  19.0       219273491   
191406          0                   8.0       233761440   

                       Vehicle Location  \
0         POINT (-122.300312 47.629782)   
1       POINT (-120.2658133 46.4063477)   
2         POINT (-122.201564 47.402358)   
3         POINT (-122.206146 47.839957)   
4         POINT (-120.611068 46.596645)   
...                                 ...   
191402  POINT (-122.2337855 47.2055969)   
191403  POINT (-122.0478543 47.1075591)   
191404    POINT (-121.903093 47.567116)   
191405    POINT (-123.807422 46.978243)   
191406  POINT (-119.2958528 46.2421419)   

                                         Electric Utility  2020 Census Tract  


In [498]:
veichels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 191407 entries, 0 to 191406
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         191407 non-null  object 
 1   County                                             191403 non-null  object 
 2   City                                               191403 non-null  object 
 3   State                                              191407 non-null  object 
 4   Postal Code                                        191403 non-null  float64
 5   Model Year                                         191407 non-null  int64  
 6   Make                                               191407 non-null  object 
 7   Model                                              191407 non-null  object 
 8   Electric Vehicle Type                              191407 non-null  object

## Initial Observations

 - `Dataset Contain 191407 and 17 columns `
 - `Some features have inopropriate data types`
 - `Some features have missing values`
 - `We need to remove some extra columns`
 - `There is no duplicate values`
 - `Model Year column data type should be changed to datetime data type`

In [500]:
# Now we check missing values
veichels.isnull().sum()

VIN (1-10)                                             0
County                                                 4
City                                                   4
State                                                  0
Postal Code                                            4
Model Year                                             0
Make                                                   0
Model                                                  0
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         0
Base MSRP                                              0
Legislative District                                 414
DOL Vehicle ID                                         0
Vehicle Location                                       8
Electric Utility                                       4
2020 Census Tract                                      4
dtype: int64

In [501]:
# Now we check duplicates
veichels.duplicated().sum()

0

## Preliminary Analysis

In [503]:
veichels.head()

VIN (1-10)     County     City State  Postal Code  Model Year    Make  \
0  5YJSA1E22K       King  Seattle    WA      98112.0        2019   TESLA   
1  3MW39FS05R     Yakima   Zillah    WA      98953.0        2024     BMW   
2  1N4AZ0CP0F       King     Kent    WA      98031.0        2015  NISSAN   
3  5YJSA1H20F  Snohomish  Bothell    WA      98012.0        2015   TESLA   
4  JTMAB3FV1N     Yakima   Yakima    WA      98908.0        2022  TOYOTA   

        Model                   Electric Vehicle Type  \
0     MODEL S          Battery Electric Vehicle (BEV)   
1        330E  Plug-in Hybrid Electric Vehicle (PHEV)   
2        LEAF          Battery Electric Vehicle (BEV)   
3     MODEL S          Battery Electric Vehicle (BEV)   
4  RAV4 PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible             270   
1             Not eligible due to low battery range              20   
2           Clean Alternative Fuel Vehicle Eligible              84   
3           Clean Alternative Fuel Vehicle Eligible             208   
4           Clean Alternative Fuel Vehicle Eligible              42   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  43.0       202233958   
1          0                  15.0       264425178   
2          0                  11.0       114962025   
3          0                   1.0       232724670   
4          0                  14.0       221023589   

                  Vehicle Location  \
0    POINT (-122.300312 47.629782)   
1  POINT (-120.2658133 46.4063477)   
2    POINT (-122.201564 47.402358)   
3    POINT (-122.206146 47.839957)   
4    POINT (-120.611068 46.596645)   

                                Electric Utility  2020 Census Tract  
0   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
1                                     PACIFICORP       5.307700e+10  
2  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
3                         PUGET SOUND ENERGY INC       5.306105e+10  
4                                     PACIFICORP       5.307700e+10

## Check for Data Types

In [505]:
veichels.dtypes

VIN (1-10)                                            object
County                                                object
City                                                  object
State                                                 object
Postal Code                                          float64
Model Year                                             int64
Make                                                  object
Model                                                 object
Electric Vehicle Type                                 object
Clean Alternative Fuel Vehicle (CAFV) Eligibility     object
Electric Range                                         int64
Base MSRP                                              int64
Legislative District                                 float64
DOL Vehicle ID                                         int64
Vehicle Location                                      object
Electric Utility                                      object
2020 Census Tract       

## Ckeck for Duplicates

In [507]:
# We check the DAta for duplicate value 
# There is no Duplicate values in our data
veichels.duplicated().sum()

0

## Removing Some Unimportant columns
`Before strating Detailed Analysis we Remove Some coluns`

In [509]:
veichels.columns

Index(['VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year',
       'Make', 'Model', 'Electric Vehicle Type',
       'Clean Alternative Fuel Vehicle (CAFV) Eligibility', 'Electric Range',
       'Base MSRP', 'Legislative District', 'DOL Vehicle ID',
       'Vehicle Location', 'Electric Utility', '2020 Census Tract'],
      dtype='object')

## Country

In [511]:
veichels.County.unique()

array(['King', 'Yakima', 'Snohomish', 'Thurston', 'Kitsap', 'Whitman',
       'Walla Walla', 'Kittitas', 'Douglas', 'Grant', 'Chelan', 'Skagit',
       'Stevens', nan, 'Maricopa', 'Clark', 'Cowlitz', 'Jefferson',
       'Island', 'Clallam', 'Klickitat', 'Spokane', 'Pierce', 'Mason',
       'San Juan', 'Benton', 'Whatcom', 'Okanogan', 'Lewis',
       'Grays Harbor', 'Asotin', 'Pacific', 'Wake', 'Franklin',
       'Skamania', 'Columbia', 'Wichita', 'Lincoln', 'El Paso',
       'Garfield', 'Adams', 'Hillsborough', 'Sonoma', 'Wahkiakum',
       'Pend Oreille', 'Marin', 'Monterey', 'Cumberland', 'Lee', 'Ferry',
       'San Diego', 'Nueces', 'Stafford', 'Washoe', 'Kern', 'DeKalb',
       'Richland', 'Pulaski', 'Platte', 'Bexar', 'Goochland', 'Kings',
       'Santa Clara', 'Madison', 'Fairfax', 'Cook', 'Los Angeles',
       'Beaufort', 'Solano', 'New York', 'Collin', 'Kootenai',
       'Contra Costa', 'Leavenworth', 'Ada', 'Anne Arundel', 'James City',
       'Alameda', 'Caddo', 'Orange', 'Pl

## City

In [513]:
veichels.City.unique()

array(['Seattle', 'Zillah', 'Kent', 'Bothell', 'Yakima', 'Redmond',
       'Yelm', 'Poulsbo', 'Olympia', 'Port Orchard', 'Shoreline',
       'Tumwater', 'Snohomish', 'Bellevue', 'Bainbridge Island',
       'Stanwood', 'Bremerton', 'Lynnwood', 'Renton', 'Monroe',
       'Lake Stevens', 'Pullman', 'Medina', 'Lacey', 'Mukilteo',
       'Suquamish', 'Silverdale', 'Kingston', 'Moxee', 'Walla Walla',
       'Vashon', 'Cle Elum', 'Everett', 'East Wenatchee', 'Woodway',
       'Quincy', 'Wenatchee', 'Moses Lake', 'Bow', 'Colville', nan,
       'Selah', 'Mount Vernon', 'Entiat', 'Anacortes', 'Enumclaw',
       'Marysville', 'Easton', 'Brier', 'Mountlake Terrace', 'Edmonds',
       'Goodyear', 'Kirkland', 'Vancouver', 'Mill Creek', 'Woodinville',
       'Tukwila', 'Mercer Island', 'Brush Prairie', 'Wapato', 'Olalla',
       'Kelso', 'Granger', 'Federal Way', 'North Bend', 'Battle Ground',
       'Mabton', 'Issaquah', 'Port Ludlow', 'Burien', 'Oak Harbor',
       'Longview', 'Sammamish', 'Lake Fo

In [514]:
veichels.City.value_counts()

City
Seattle                    31436
Bellevue                    9582
Redmond                     6883
Vancouver                   6744
Bothell                     6278
                           ...  
Holloman Air Force Base        1
Rosamond                       1
Newbury Park                   1
Saratoga Springs               1
Fremont                        1
Name: count, Length: 744, dtype: int64

In [515]:
veichels.City.dtypes

dtype('O')

## State

In [517]:
veichels.State.unique()

array(['WA', 'BC', 'AE', 'AZ', 'NC', 'NV', 'TX', 'CO', 'FL', 'CA', 'PA',
       'AL', 'VA', 'GA', 'SC', 'MO', 'NY', 'IL', 'NE', 'ID', 'KS', 'MD',
       'LA', 'DE', 'OH', 'AR', 'WY', 'UT', 'MN', 'HI', 'NM', 'OR', 'IN',
       'MI', 'NJ', 'RI', 'NH', 'CT', 'IA', 'MA', 'OK', 'KY', 'DC', 'AK',
       'MT', 'WI'], dtype=object)

In [518]:
(
    veichels.State.replace({
        'WA' :'Washington',
        'BC' :'British Columbia',
        'AE' :'Africa',
        'AZ' : 'Arizona',
        'NC' : 'North Carolina',
        'NV' : 'Nevada',
        'TX' : 'Taxas',
        'CO' : 'Colorado',
        'FL' :'Florida',
        'CA' : 'California',
        'PA' : 'Pennsylvania',
        'AL' : 'Alabama',
        'VA' : 'Virginia',
        'GA' : 'Georgia',
        'SC' : 'South Carolina',
        'MO' : 'Missouri',
        'NY' : 'New York',
        'IL' : 'ILLINOIS',
         'NE': 'Nebraska',
        'ID': 'Idaho',
        'KS': 'Kansas',
        'MD': 'Maryland',
        'LA': 'Louisiana',
        'DE': 'Delaware',
        'OH': 'Ohio',
        'AR': 'Arkansas',
        'WY': 'Wyoming',
        'UT': 'Utah',
        'MN': 'Minnesota',
        'HI': 'Hawaii',
        'NM': 'New Mexico',
        'OR': 'Oregon',
        'IN': 'Indiana',
        'MI': 'Michigan',
        'NJ': 'New Jersey',
        'RI': 'Rhode Island',
        'NH': 'New Hampshire',
        'CT': 'Connecticut',
        'IA': 'Iowa',
        'MA': 'Massachusetts',
        'OK': 'Oklahoma',
        'KY': 'Kentucky',
        'DC': 'District of Columbia',
        'AK': 'Alaska',
        'MT': 'Montana',
        'WI': 'Wisconsin'
    })
)

0         Washington
1         Washington
2         Washington
3         Washington
4         Washington
             ...    
191402    Washington
191403    Washington
191404    Washington
191405    Washington
191406    Washington
Name: State, Length: 191407, dtype: object

In [519]:
veichels.State.value_counts()

State
WA    190993
CA       107
VA        53
MD        32
TX        25
CO        16
NC        15
IL        14
AZ        13
NV         9
FL         9
NY         8
NJ         8
HI         8
OR         7
CT         7
AL         7
SC         7
UT         6
GA         6
MO         5
ID         4
OH         4
LA         4
DC         3
IN         3
PA         3
KY         3
KS         3
MA         3
BC         3
AR         2
MI         2
RI         2
NE         2
AK         1
MT         1
OK         1
DE         1
IA         1
NH         1
NM         1
MN         1
WY         1
AE         1
WI         1
Name: count, dtype: int64

## Postal Code

In [521]:
# we convert postal code column into interger data type
pd.to_numeric(veichels['Postal Code'], errors='coerce').astype('Int64')

0         98112
1         98953
2         98031
3         98012
4         98908
          ...  
191402    98390
191403    98396
191404    98024
191405    98520
191406    99352
Name: Postal Code, Length: 191407, dtype: Int64

In [522]:
(
    veichels['Postal Code']
    .pipe(lambda ser : pd.to_numeric(ser,errors='coerce').astype('Int64'))
)

0         98112
1         98953
2         98031
3         98012
4         98908
          ...  
191402    98390
191403    98396
191404    98024
191405    98520
191406    99352
Name: Postal Code, Length: 191407, dtype: Int64

## Model Year

In [524]:
# we convert the model year column into datetime data type 
pd.to_datetime(veichels['Model Year'],format='%Y').dt.year

0         2019
1         2024
2         2015
3         2015
4         2022
          ... 
191402    2019
191403    2023
191404    2024
191405    2022
191406    2023
Name: Model Year, Length: 191407, dtype: int32

## Make

In [526]:
veichels.Make.unique()

array(['TESLA', 'BMW', 'NISSAN', 'TOYOTA', 'FORD', 'FIAT', 'CHRYSLER',
       'AUDI', 'VOLVO', 'PORSCHE', 'JEEP', 'HYUNDAI', 'VOLKSWAGEN',
       'CHEVROLET', 'KIA', 'MERCEDES-BENZ', 'MINI', 'MITSUBISHI', 'HONDA',
       'RIVIAN', 'JAGUAR', 'DODGE', 'CADILLAC', 'LUCID', 'SUBARU',
       'POLESTAR', 'LEXUS', 'SMART', 'ALFA ROMEO', 'MAZDA', 'FISKER',
       'LINCOLN', 'GENESIS', 'LAND ROVER', 'GMC', 'TH!NK',
       'AZURE DYNAMICS', 'WHEEGO ELECTRIC CARS', 'BENTLEY', 'ROLLS ROYCE',
       'ACURA', 'RAM'], dtype=object)

In [527]:
veichels.Make.replace('TH!NK','THINK').unique()

array(['TESLA', 'BMW', 'NISSAN', 'TOYOTA', 'FORD', 'FIAT', 'CHRYSLER',
       'AUDI', 'VOLVO', 'PORSCHE', 'JEEP', 'HYUNDAI', 'VOLKSWAGEN',
       'CHEVROLET', 'KIA', 'MERCEDES-BENZ', 'MINI', 'MITSUBISHI', 'HONDA',
       'RIVIAN', 'JAGUAR', 'DODGE', 'CADILLAC', 'LUCID', 'SUBARU',
       'POLESTAR', 'LEXUS', 'SMART', 'ALFA ROMEO', 'MAZDA', 'FISKER',
       'LINCOLN', 'GENESIS', 'LAND ROVER', 'GMC', 'THINK',
       'AZURE DYNAMICS', 'WHEEGO ELECTRIC CARS', 'BENTLEY', 'ROLLS ROYCE',
       'ACURA', 'RAM'], dtype=object)

In [528]:
veichels.Make.value_counts()

Make
TESLA                   84670
CHEVROLET               14321
NISSAN                  14239
FORD                    10048
KIA                      8671
BMW                      7993
TOYOTA                   7181
VOLKSWAGEN               5408
JEEP                     4996
HYUNDAI                  4977
RIVIAN                   4751
VOLVO                    4469
AUDI                     3890
CHRYSLER                 3393
MERCEDES-BENZ            1784
PORSCHE                  1182
MITSUBISHI                996
MINI                      984
SUBARU                    970
POLESTAR                  939
HONDA                     857
FIAT                      781
DODGE                     619
MAZDA                     581
LEXUS                     567
CADILLAC                  552
LINCOLN                   281
LUCID                     263
SMART                     259
JAGUAR                    234
GENESIS                   219
FISKER                    181
LAND ROVER                 60
ALFA 

## Model

In [530]:
veichels.Model.unique()

array(['MODEL S', '330E', 'LEAF', 'RAV4 PRIME', 'MODEL 3', 'FUSION',
       'PRIUS PRIME', '500', 'PACIFICA', 'I4', 'MODEL Y', 'PRIUS PLUG-IN',
       'A3', 'I3', 'X5', 'XC90', 'TAYCAN', 'WRANGLER', 'KONA', 'XC60',
       'MODEL X', 'F-150', 'ID.4', 'VOLT', 'SPORTAGE', 'GLC-CLASS',
       'SORENTO', 'RAV4', 'COUNTRYMAN', 'C-MAX', 'HARDTOP', 'E-TRON',
       'BOLT EV', 'OUTLANDER', 'X3', 'CLARITY', 'SOUL EV', 'TUCSON',
       '745E', 'NIRO', 'R1S', 'Q5 E', 'B-CLASS', 'I-PACE', 'C40',
       'KONA ELECTRIC', 'EV6', 'BZ4X', 'GRAND CHEROKEE', 'Q8', 'CAYENNE',
       'HORNET', 'BOLT EUV', 'MUSTANG MACH-E', 'EQS-CLASS SEDAN', '530E',
       'IONIQ 5', 'SPARK', 'S60', 'LYRIQ', 'AIR', 'EV9', 'TRANSIT',
       'CYBERTRUCK', 'S-CLASS', 'SOLTERRA', 'IX', 'E-GOLF', 'XC40', 'PS2',
       'Q4', 'EQS-CLASS SUV', 'IONIQ 6', 'IONIQ', 'EQB-CLASS',
       'RS E-TRON GT', 'ARIYA', 'R1T', 'ROADSTER', 'OPTIMA', 'SOUL',
       'ESCAPE', 'NX', 'I7', 'I8', 'EQ FORTWO', 'E-TRON SPORTBACK',
       'C-CLASS', 'TO

In [531]:
veichels.Model.str.strip().unique()

array(['MODEL S', '330E', 'LEAF', 'RAV4 PRIME', 'MODEL 3', 'FUSION',
       'PRIUS PRIME', '500', 'PACIFICA', 'I4', 'MODEL Y', 'PRIUS PLUG-IN',
       'A3', 'I3', 'X5', 'XC90', 'TAYCAN', 'WRANGLER', 'KONA', 'XC60',
       'MODEL X', 'F-150', 'ID.4', 'VOLT', 'SPORTAGE', 'GLC-CLASS',
       'SORENTO', 'RAV4', 'COUNTRYMAN', 'C-MAX', 'HARDTOP', 'E-TRON',
       'BOLT EV', 'OUTLANDER', 'X3', 'CLARITY', 'SOUL EV', 'TUCSON',
       '745E', 'NIRO', 'R1S', 'Q5 E', 'B-CLASS', 'I-PACE', 'C40',
       'KONA ELECTRIC', 'EV6', 'BZ4X', 'GRAND CHEROKEE', 'Q8', 'CAYENNE',
       'HORNET', 'BOLT EUV', 'MUSTANG MACH-E', 'EQS-CLASS SEDAN', '530E',
       'IONIQ 5', 'SPARK', 'S60', 'LYRIQ', 'AIR', 'EV9', 'TRANSIT',
       'CYBERTRUCK', 'S-CLASS', 'SOLTERRA', 'IX', 'E-GOLF', 'XC40', 'PS2',
       'Q4', 'EQS-CLASS SUV', 'IONIQ 6', 'IONIQ', 'EQB-CLASS',
       'RS E-TRON GT', 'ARIYA', 'R1T', 'ROADSTER', 'OPTIMA', 'SOUL',
       'ESCAPE', 'NX', 'I7', 'I8', 'EQ FORTWO', 'E-TRON SPORTBACK',
       'C-CLASS', 'TO

In [532]:
veichels.Model.pipe(lambda ser : ser.str.replace('-','')).replace('MODEL S','MODEL').unique()

array(['MODEL', '330E', 'LEAF', 'RAV4 PRIME', 'MODEL 3', 'FUSION',
       'PRIUS PRIME', '500', 'PACIFICA', 'I4', 'MODEL Y', 'PRIUS PLUGIN',
       'A3', 'I3', 'X5', 'XC90', 'TAYCAN', 'WRANGLER', 'KONA', 'XC60',
       'MODEL X', 'F150', 'ID.4', 'VOLT', 'SPORTAGE', 'GLCCLASS',
       'SORENTO', 'RAV4', 'COUNTRYMAN', 'CMAX', 'HARDTOP', 'ETRON',
       'BOLT EV', 'OUTLANDER', 'X3', 'CLARITY', 'SOUL EV', 'TUCSON',
       '745E', 'NIRO', 'R1S', 'Q5 E', 'BCLASS', 'IPACE', 'C40',
       'KONA ELECTRIC', 'EV6', 'BZ4X', 'GRAND CHEROKEE', 'Q8', 'CAYENNE',
       'HORNET', 'BOLT EUV', 'MUSTANG MACHE', 'EQSCLASS SEDAN', '530E',
       'IONIQ 5', 'SPARK', 'S60', 'LYRIQ', 'AIR', 'EV9', 'TRANSIT',
       'CYBERTRUCK', 'SCLASS', 'SOLTERRA', 'IX', 'EGOLF', 'XC40', 'PS2',
       'Q4', 'EQSCLASS SUV', 'IONIQ 6', 'IONIQ', 'EQBCLASS',
       'RS ETRON GT', 'ARIYA', 'R1T', 'ROADSTER', 'OPTIMA', 'SOUL',
       'ESCAPE', 'NX', 'I7', 'I8', 'EQ FORTWO', 'ETRON SPORTBACK',
       'CCLASS', 'TONALE', 'CX90', 'EQ

In [533]:
veichels.Model.str.replace('MODEL S','MODEL').unique()

array(['MODEL', '330E', 'LEAF', 'RAV4 PRIME', 'MODEL 3', 'FUSION',
       'PRIUS PRIME', '500', 'PACIFICA', 'I4', 'MODEL Y', 'PRIUS PLUG-IN',
       'A3', 'I3', 'X5', 'XC90', 'TAYCAN', 'WRANGLER', 'KONA', 'XC60',
       'MODEL X', 'F-150', 'ID.4', 'VOLT', 'SPORTAGE', 'GLC-CLASS',
       'SORENTO', 'RAV4', 'COUNTRYMAN', 'C-MAX', 'HARDTOP', 'E-TRON',
       'BOLT EV', 'OUTLANDER', 'X3', 'CLARITY', 'SOUL EV', 'TUCSON',
       '745E', 'NIRO', 'R1S', 'Q5 E', 'B-CLASS', 'I-PACE', 'C40',
       'KONA ELECTRIC', 'EV6', 'BZ4X', 'GRAND CHEROKEE', 'Q8', 'CAYENNE',
       'HORNET', 'BOLT EUV', 'MUSTANG MACH-E', 'EQS-CLASS SEDAN', '530E',
       'IONIQ 5', 'SPARK', 'S60', 'LYRIQ', 'AIR', 'EV9', 'TRANSIT',
       'CYBERTRUCK', 'S-CLASS', 'SOLTERRA', 'IX', 'E-GOLF', 'XC40', 'PS2',
       'Q4', 'EQS-CLASS SUV', 'IONIQ 6', 'IONIQ', 'EQB-CLASS',
       'RS E-TRON GT', 'ARIYA', 'R1T', 'ROADSTER', 'OPTIMA', 'SOUL',
       'ESCAPE', 'NX', 'I7', 'I8', 'EQ FORTWO', 'E-TRON SPORTBACK',
       'C-CLASS', 'TONA

## Electric Vehicle Type

In [535]:
veichels['Electric Vehicle Type']

0                 Battery Electric Vehicle (BEV)
1         Plug-in Hybrid Electric Vehicle (PHEV)
2                 Battery Electric Vehicle (BEV)
3                 Battery Electric Vehicle (BEV)
4         Plug-in Hybrid Electric Vehicle (PHEV)
                           ...                  
191402            Battery Electric Vehicle (BEV)
191403            Battery Electric Vehicle (BEV)
191404            Battery Electric Vehicle (BEV)
191405            Battery Electric Vehicle (BEV)
191406            Battery Electric Vehicle (BEV)
Name: Electric Vehicle Type, Length: 191407, dtype: object

In [536]:
veichels['Electric Vehicle Type'].unique()

array(['Battery Electric Vehicle (BEV)',
       'Plug-in Hybrid Electric Vehicle (PHEV)'], dtype=object)

In [537]:
veichels['Electric Vehicle Type'].value_counts()

Electric Vehicle Type
Battery Electric Vehicle (BEV)            149687
Plug-in Hybrid Electric Vehicle (PHEV)     41720
Name: count, dtype: int64

In [538]:
(
    veichels['Electric Vehicle Type']
    .replace({
        'Battery Electric Vehicle (BEV)' : 'BEV',
        'Plug-in Hybrid Electric Vehicle (PHEV)' : 'PHEV'
    })
)

0          BEV
1         PHEV
2          BEV
3          BEV
4         PHEV
          ... 
191402     BEV
191403     BEV
191404     BEV
191405     BEV
191406     BEV
Name: Electric Vehicle Type, Length: 191407, dtype: object

## Clean Alternative Fuel Vehicle (CAFV) Eligibility

In [540]:
veichels['Clean Alternative Fuel Vehicle (CAFV) Eligibility']

0                   Clean Alternative Fuel Vehicle Eligible
1                     Not eligible due to low battery range
2                   Clean Alternative Fuel Vehicle Eligible
3                   Clean Alternative Fuel Vehicle Eligible
4                   Clean Alternative Fuel Vehicle Eligible
                                ...                        
191402              Clean Alternative Fuel Vehicle Eligible
191403    Eligibility unknown as battery range has not b...
191404    Eligibility unknown as battery range has not b...
191405    Eligibility unknown as battery range has not b...
191406    Eligibility unknown as battery range has not b...
Name: Clean Alternative Fuel Vehicle (CAFV) Eligibility, Length: 191407, dtype: object

In [541]:
veichels['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].unique()

array(['Clean Alternative Fuel Vehicle Eligible',
       'Not eligible due to low battery range',
       'Eligibility unknown as battery range has not been researched'],
      dtype=object)

In [542]:
veichels['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].value_counts()

Clean Alternative Fuel Vehicle (CAFV) Eligibility
Eligibility unknown as battery range has not been researched    102133
Clean Alternative Fuel Vehicle Eligible                          68785
Not eligible due to low battery range                            20489
Name: count, dtype: int64

In [543]:
(
    veichels['Clean Alternative Fuel Vehicle (CAFV) Eligibility']
    .replace({
        'Eligibility unknown as battery range has not been researched': 'unknown',
        'Clean Alternative Fuel Vehicle Eligible' : 'eligible',
        'Not eligible due to low battery range' : 'not eligible'
    })
)

0             eligible
1         not eligible
2             eligible
3             eligible
4             eligible
              ...     
191402        eligible
191403         unknown
191404         unknown
191405         unknown
191406         unknown
Name: Clean Alternative Fuel Vehicle (CAFV) Eligibility, Length: 191407, dtype: object

## Electric Range

In [545]:
veichels['Electric Range']

0         270
1          20
2          84
3         208
4          42
         ... 
191402    220
191403      0
191404      0
191405      0
191406      0
Name: Electric Range, Length: 191407, dtype: int64

In [546]:
veichels['Electric Range'].dtypes

dtype('int64')

In [547]:
veichels['Electric Range'].isnull().sum()

0

## Base MSRP

In [549]:
veichels['Base MSRP']

0         0
1         0
2         0
3         0
4         0
         ..
191402    0
191403    0
191404    0
191405    0
191406    0
Name: Base MSRP, Length: 191407, dtype: int64

In [550]:
veichels['Base MSRP'].dtypes

dtype('int64')

In [551]:
veichels['Base MSRP'].isnull().sum()

0

In [552]:
veichels['Legislative District']

0         43.0
1         15.0
2         11.0
3          1.0
4         14.0
          ... 
191402    31.0
191403    31.0
191404     5.0
191405    19.0
191406     8.0
Name: Legislative District, Length: 191407, dtype: float64

## Legislative District

In [554]:
veichels['Legislative District']

0         43.0
1         15.0
2         11.0
3          1.0
4         14.0
          ... 
191402    31.0
191403    31.0
191404     5.0
191405    19.0
191406     8.0
Name: Legislative District, Length: 191407, dtype: float64

In [555]:
veichels['Legislative District'].astype('Int64').unique()

<IntegerArray>
[  43,   15,   11,    1,   14,   48,    2,   23,   45,   22,   35,   47,   32,
   21,   46,   10,   26,   39,   44,    9,   34,   37,   16,   13,   12,   40,
    7, <NA>,   31,   38,   30,   36,   17,   49,   41,   33,   19,    5,   18,
   24,   20,   29,    3,   28,    4,    6,   27,   25,    8,   42]
Length: 50, dtype: Int64

In [556]:
veichels['Electric Utility'].unique()

array(['CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)', 'PACIFICORP',
       'PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)',
       'PUGET SOUND ENERGY INC', 'AVISTA CORP',
       'PUD NO 1 OF DOUGLAS COUNTY', 'PUD NO 2 OF GRANT COUNTY',
       'PUD NO 1 OF CHELAN COUNTY', nan,
       'NON WASHINGTON STATE ELECTRIC UTILITY',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF CLARK COUNTY - (WA)',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF COWLITZ COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PACIFICORP||BENTON RURAL ELECTRIC ASSN',
       'BONNEVILLE POWER ADMINISTRATION||PUGET SOUND ENERGY INC||PUD NO 1 OF JEFFERSON COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PUD 1 OF SNOHOMISH COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PACIFICORP||PUD NO 1 OF CLARK COUNTY - (WA)',
       'NO KNOWN ELECTRIC UTILITY SERVICE',
       'BONNEVILLE POWER ADMINISTRATION||CITY OF PORT ANGELES - (WA)',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF MASON COUNTY|PUD NO 1 OF J

In [557]:
veichels['Electric Utility'].value_counts()

Electric Utility
PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)                                                                 70546
PUGET SOUND ENERGY INC                                                                                        38778
CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)                                                                  33567
BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF CLARK COUNTY - (WA)                                              11067
BONNEVILLE POWER ADMINISTRATION||CITY OF TACOMA - (WA)||PENINSULA LIGHT COMPANY                                8446
                                                                                                              ...  
BONNEVILLE POWER ADMINISTRATION||CITY OF TACOMA - (WA)||ALDER MUTUAL LIGHT CO, INC|PENINSULA LIGHT COMPANY        2
CITY OF SEATTLE - (WA)                                                                                            2
BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF CLALLAM CO

## Cleaning Operation

In [669]:
def clean_data(df):
    return(
        df.assign(**{
            col :df[col].str.strip()
            for col in df.select_dtypes(include = "O").columns
        }) # assign end
        # In this line we covert the all columns into lower case
        .rename(columns = str.lower)
        # In this line we rename the all columns
        .rename(columns = {
            'vin (1-10)': 'vin',
            'county' :'country',
            'make' : 'manufacturer',
            'electric vehicle type' :'electric_type',
            'clean alternative fuel vehicle (cafv) eligibility':'clean_fuel_eligibility',
            'electric range':'electric_range',
            'base msrp':'retail_price',
            'legislative district':'legislative_district',
            'dol vehicle id' :'vehicle_id',
            'vehicle location':'vehicle_location',
            'electric utility':'electric_utility',
            '2020 census tract':'2020_census_tract',
            'postal code': 'postal_code',
            'model year': 'model_year'
        })
        .assign(
            # we replace the all values with complete name of state column
            state = lambda df_ :(
                df_
                .state
                .replace({
                        'WA' :'Washington',
                        'BC' :'British Columbia',
                        'AE' :'Africa',
                        'AZ' : 'Arizona',
                        'NC' : 'North Carolina',
                        'NV' : 'Nevada',
                        'TX' : 'Taxas',
                        'CO' : 'Colorado',
                        'FL' :'Florida',
                        'CA' : 'California',
                        'PA' : 'Pennsylvania',
                        'AL' : 'Alabama',
                        'VA' : 'Virginia',
                        'GA' : 'Georgia',
                        'SC' : 'South Carolina',
                        'MO' : 'Missouri',
                        'NY' : 'New York',
                        'IL' : 'ILLINOIS',
                         'NE': 'Nebraska',
                        'ID': 'Idaho',
                        'KS': 'Kansas',
                        'MD': 'Maryland',
                        'LA': 'Louisiana',
                        'DE': 'Delaware',
                        'OH': 'Ohio',
                        'AR': 'Arkansas',
                        'WY': 'Wyoming',
                        'UT': 'Utah',
                        'MN': 'Minnesota',
                        'HI': 'Hawaii',
                        'NM': 'New Mexico',
                        'OR': 'Oregon',
                        'IN': 'Indiana',
                        'MI': 'Michigan',
                        'NJ': 'New Jersey',
                        'RI': 'Rhode Island',
                        'NH': 'New Hampshire',
                        'CT': 'Connecticut',
                        'IA': 'Iowa',
                        'MA': 'Massachusetts',
                        'OK': 'Oklahoma',
                        'KY': 'Kentucky',
                        'DC': 'District of Columbia',
                        'AK': 'Alaska',
                        'MT': 'Montana',
                        'WI': 'Wisconsin'
    })
            ), #state column end
            
            # we convert the postal_code column into int data type
            postal_code = lambda df_ : (
                df_
                .postal_code
                .pipe(lambda ser : pd.to_numeric(ser,errors='coerce').astype('Int64'))
                ),  #posatl code column end
            
            # model_year column convert into datetime data type
            model_year = lambda df_ : pd.to_datetime(df_.model_year).dt.year,
            
            # manufacturer column
            manufacturer = lambda df_ : df_.manufacturer.replace('TH!NK','THINK'),

            # model column
            # we remove the - from  model columnabs
            model = lambda df_ : (
                df_
                .model
                .pipe(lambda ser : ser.str.replace('-',''))
            ),    # model column end

            # electric_type column
            electric_type = lambda df_ :(
                df_
                .electric_type
                .replace({
                    'Battery Electric Vehicle (BEV)' : 'BEV',
                    'Plug-in Hybrid Electric Vehicle (PHEV)' : 'PHEV'
                    
                })
            ),    #electric_type column end
            
            # clean_fuel_eligibility column
            clean_fuel_eligibility = lambda df_ :(          #clean_fuel_eligibility strat
                df_
                .clean_fuel_eligibility
                .replace({
                    'Eligibility unknown as battery range has not been researched': 'unknown',
                    'Clean Alternative Fuel Vehicle Eligible' : 'eligible',
                    'Not eligible due to low battery range' : 'not eligible'
                        })
            ), 

            # Registered district column
            legislative_district = lambda df_:(
                df_
                .legislative_district
                .astype('Int64')
            ),    # registred_district end
            
            # vehicle_location column
            vehicle_location = lambda df_ : df_.vehicle_location,

            # electric_range column
            # we move the electric_range cloumn ate end of datafraem because it is our tagrget variable
            electric_range = lambda df_ : df_.pop('electric_range')
        )
        we drop some unnecessary 
        .drop(columns = ['vin','postal_code','vehicle_id','retail_price','electric_utility','2020_census_tract','legislative_district'])
         
    )
    
     


In [676]:
vehicles_cleaned = clean_data(veichels)
vehicles_cleaned

country       city       state  model_year manufacturer  \
0               King    Seattle  Washington        1970        TESLA   
1             Yakima     Zillah  Washington        1970          BMW   
2               King       Kent  Washington        1970       NISSAN   
3          Snohomish    Bothell  Washington        1970        TESLA   
4             Yakima     Yakima  Washington        1970       TOYOTA   
...              ...        ...         ...         ...          ...   
191402        Pierce     Sumner  Washington        1970        TESLA   
191403        Pierce   Wilkeson  Washington        1970       NISSAN   
191404          King  Fall City  Washington        1970        TESLA   
191405  Grays Harbor   Aberdeen  Washington        1970        TESLA   
191406        Benton   Richland  Washington        1970        TESLA   

             model electric_type clean_fuel_eligibility  \
0          MODEL S           BEV               eligible   
1             330E          PHEV           not eligible   
2             LEAF           BEV               eligible   
3          MODEL S           BEV               eligible   
4       RAV4 PRIME          PHEV               eligible   
...            ...           ...                    ...   
191402     MODEL 3           BEV               eligible   
191403       ARIYA           BEV                unknown   
191404     MODEL Y           BEV                unknown   
191405     MODEL X           BEV                unknown   
191406     MODEL Y           BEV                unknown   

                       vehicle_location  electric_range  
0         POINT (-122.300312 47.629782)             270  
1       POINT (-120.2658133 46.4063477)              20  
2         POINT (-122.201564 47.402358)              84  
3         POINT (-122.206146 47.839957)             208  
4         POINT (-120.611068 46.596645)              42  
...                                 ...             ...  
191402  POINT (-122.2337855 47.2055969)             220  
191403  POINT (-122.0478543 47.1075591)               0  
191404    POINT (-121.903093 47.567116)               0  
191405    POINT (-123.807422 46.978243)               0  
191406  POINT (-119.2958528 46.2421419)               0  

[191407 rows x 10 columns]

## Split the Data

In [695]:
x = vehicles_cleaned.drop(columns = 'electric_range')  
y = vehicles_cleaned.electric_range.copy()

In [703]:
X_,X_test,y_,y_test = train_test_split(x,y,test_size=.2,random_state=42)

X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size=.2,random_state=42)

In [707]:
print(X_train.shape,y_train.shape)
print(X_test.shape,y_test.shape)
print(X_val.shape,y_val.shape)

(122500, 9) (122500,)
(38282, 9) (38282,)
(30625, 9) (30625,)


## Export the data Subsets

In [714]:
def export_data(x,y,name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    x.join(y).to_csv(file_path,index=False)
    return pd.read_csv(file_path).head()
    

In [716]:
export_data(X_train,y_train,'train')

country       city       state  model_year manufacturer     model  \
0  Snohomish    Edmonds  Washington        1970    CHEVROLET  BOLT EUV   
1       King   Kirkland  Washington        1970       NISSAN     ARIYA   
2       King    Kenmore  Washington        1970        TESLA   MODEL Y   
3       King  Sammamish  Washington        1970        TESLA   MODEL S   
4     Pierce     Tacoma  Washington        1970        TESLA   MODEL Y   

  electric_type clean_fuel_eligibility                 vehicle_location  \
0           BEV                unknown   POINT (-122.3757836 47.800042)   
1           BEV                unknown  POINT (-122.2026532 47.7210518)   
2           BEV                unknown    POINT (-122.2478069 47.75817)   
3           BEV               eligible  POINT (-122.0181135 47.5880568)   
4           BEV                unknown     POINT (-122.48056 47.214139)   

   electric_range  
0               0  
1               0  
2               0  
3             210  
4               0

In [718]:
export_data(X_test,y_test,'test')

country          city       state  model_year manufacturer    model  \
0  Snohomish      Lynnwood  Washington        1970   VOLKSWAGEN     ID.4   
1       King        Vashon  Washington        1970       NISSAN     LEAF   
2  Snohomish  Lake Stevens  Washington        1970        VOLVO      C40   
3       King        Renton  Washington        1970      HYUNDAI  IONIQ 5   
4  Snohomish       Bothell  Washington        1970        TESLA  MODEL Y   

  electric_type clean_fuel_eligibility                 vehicle_location  \
0           BEV                unknown    POINT (-122.287143 47.812199)   
1           BEV               eligible    POINT (-122.466938 47.429244)   
2           BEV                unknown  POINT (-122.0816912 48.0122934)   
3           BEV                unknown    POINT (-122.134502 47.492627)   
4           BEV                unknown    POINT (-122.206146 47.839957)   

   electric_range  
0               0  
1             151  
2               0  
3               0  
4               0

In [720]:
export_data(X_val,y_val,'val')

country        city       state  model_year manufacturer    model  \
0     King    Kirkland  Washington        1970        TESLA  MODEL Y   
1     King     Seattle  Washington        1970         FIAT      500   
2  Whatcom  Bellingham  Washington        1970        TESLA  MODEL S   
3     King    Kirkland  Washington        1970       NISSAN     LEAF   
4     King        Kent  Washington        1970        TESLA  MODEL 3   

  electric_type clean_fuel_eligibility                 vehicle_location  \
0           BEV                unknown   POINT (-122.1925969 47.676241)   
1           BEV               eligible  POINT (-122.2753318 47.5411053)   
2           BEV               eligible     POINT (-122.464218 48.79135)   
3           BEV               eligible  POINT (-122.2026532 47.7210518)   
4           BEV                unknown  POINT (-122.2612577 47.3821912)   

   electric_range  
0               0  
1              84  
2             330  
3             107  
4               0